From 31a6014321d5c04c5a5148b1c50c7385fffe0bf0 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Mon, 23 Sep 2019 12:13:41 +0200
Subject: [PATCH 09/11] Pinmux WIFI or SD based on card detect GPIO input
 status

Pinmux the SD card by default, and if the SD card detect line is high (inactive)
then pinmux the SD1 interface to EMIO instead.
---
 .../psu_init_gpl.c                            | 51 +++++++++----------
 1 file changed, 25 insertions(+), 26 deletions(-)

diff --git a/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c b/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
index c770474628..4e119d7bd0 100644
--- a/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
+++ b/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
@@ -12925,7 +12925,7 @@ unsigned long psu_mio_init_data(void)
     * Configures MIO Pin 45 peripheral interface mapping
     * (OFFSET, MASK, VALUE)      (0XFF1800B4, 0x000000FEU ,0x00000010U)
     */
-	PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000010U);
+	PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000000U);
 /*##################################################################### */
 
     /*
@@ -22138,6 +22138,28 @@ int psu_protection(void)
 	return 0;
 }
 
+void psu_init_sdio_pinmux(void)
+{
+	/* Read GPIO bank 1 register to get GPIO 45 input state (SD card detect) */
+	if (Xil_In32(0xFF0A0064U) & (1U << 19)) {
+		/* SD not detected, pinmux the WiFi chip card */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_46_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_47_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_48_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_49_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_50_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_51_OFFSET, 0x000000FEU, 0x00000000U);
+		/* Set feedback clock to EMIO */
+		PSU_Mask_Write(IOU_SLCR_SDIO_CLK_CTRL_OFFSET,0x00040000U, 0x00040000U);
+		/* Voltage support */
+		PSU_Mask_Write(IOU_SLCR_SD_CONFIG_REG2_OFFSET, 0x33843384U, 0x02801284U);
+	} else {
+		/* Set pin 45 as SD card detect */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000010U);
+	}
+}
+
 int
 psu_init(void)
 {
@@ -22151,39 +22173,16 @@ psu_init(void)
 	status &=  psu_ddr_phybringup_data();
 	status &=  psu_peripherals_init_data();
 	status &=  init_serdes();
-	init_peripheral();
-
-	status &=  psu_peripherals_powerdwn_data();
-	status &=    psu_afi_config();
-	psu_ddr_qos_init_data();
-
-	if (status == 0)
-		return 1;
-	return 0;
-}
 
-int psu_init_ddr_self_refresh(void) {
-
-    int status = 1;
-
-	status &= psu_mio_init_data();
-	status &=  psu_peripherals_pre_init_data();
-	status &=   psu_pll_init_data();
-	status &=   psu_clock_init_data();
-	status &=  psu_ddr_init_data();
-	status &=  psu_peripherals_init_data();
-	status &=  init_serdes();
 	init_peripheral();
 
 	status &=  psu_peripherals_powerdwn_data();
 	status &=    psu_afi_config();
 	psu_ddr_qos_init_data();
 
+    psu_init_sdio_pinmux();
+
 	if (status == 0)
 		return 1;
 	return 0;
-
-
 }
-
-
-- 
2.17.1

